src/backend/distributed/master/master_truncate.c

```diff
@@ -65,22 +67,79 @@ citus_truncate_trigger(PG_FUNCTION_ARGS)
 
 	if (partitionMethod == DISTRIBUTE_BY_APPEND)
 	{
+		Oid schemaId = get_rel_namespace(relationId);
+		char *schemaName = get_namespace_name(schemaId);
+		char *relationName = get_rel_name(relationId);
+
 		DirectFunctionCall3(master_drop_all_shards,
 							ObjectIdGetDatum(relationId),
 							CStringGetTextDatum(relationName),
 							CStringGetTextDatum(schemaName));
 	}
 	else
 	{
-		StringInfo truncateStatement = makeStringInfo();
-		char *qualifiedTableName = quote_qualified_identifier(schemaName, relationName);
+		List *taskList = TruncateTaskList(relationId);
 
-		appendStringInfo(truncateStatement, "TRUNCATE TABLE %s CASCADE",
-						 qualifiedTableName);
-
-		DirectFunctionCall1(master_modify_multiple_shards,
-							CStringGetTextDatum(truncateStatement->data));
+		if (MultiShardConnectionType == SEQUENTIAL_CONNECTION)
+		{
+			ExecuteModifyTasksSequentiallyWithoutResults(taskList, CMD_UTILITY);
+		}
+		else
+		{
+			ExecuteModifyTasksWithoutResults(taskList);
+		}
 	}
 
 	PG_RETURN_DATUM(PointerGetDatum(NULL));
 }
+
+
+/*
+ * TruncateTaskList returns a list of tasks to execute a TRUNCATE on a
+ * distributed table. This is handled separately from other DDL commands
+ * because we handle it via the TRUNCATE trigger, which is called whenever
+ * a truncate cascades.
+ */
+static List *
+TruncateTaskList(Oid relationId)
+{
+	List *shardIntervalList = LoadShardIntervalList(relationId);
+	ListCell *shardIntervalCell = NULL;
+	List *taskList = NIL;
+	int taskId = 1;
+
+	Oid schemaId = get_rel_namespace(relationId);
+	char *schemaName = get_namespace_name(schemaId);
+	char *relationName = get_rel_name(relationId);
+
+	/* lock metadata before getting placement lists */
+	LockShardListMetadata(shardIntervalList, ShareLock);
+
+	foreach(shardIntervalCell, shardIntervalList)
+	{
+		ShardInterval *shardInterval = (ShardInterval *) lfirst(shardIntervalCell);
+		uint64 shardId = shardInterval->shardId;
+		StringInfo shardQueryString = makeStringInfo();
+		Task *task = NULL;
+		char *shardName = pstrdup(relationName);
+
+		AppendShardIdToName(&shardName, shardId);
+
+		appendStringInfo(shardQueryString, "TRUNCATE TABLE %s CASCADE",
```

Comment:
Very minor, not related to this PR, feel free to ignore: I've realized that we don't have a TRUNCATE test on schemas with quotes. You might consider adding one TRUNCATE test on one of [these tables](https://github.com/citusdata/citus/blob/master/src/test/regress/sql/multi_schema_support.sql#L793-L799)

Reply:
added